### DISCUSSION TITLE
Limiting scope of a PAT to a single repository

### DISCUSSION THREAD

**Discussion by ghost** [community/community#21999] [New to GitHub]
<p>I’m currently using TravisCI in our team to build private NPM packages and publish them to Github packages. I have created a bot account, added it to our team and using its PAT in Travis. Now let’s say John has access to repository A but doesn’t have access to repository B. Problem is because of the bot account PAT having access to both repositories, John can edit “.travis.yml” file in repository A and therefore gaining access to repository B. My first thought was to create a bot account per repository, Unfortunately that is not going to work because I need to pay $9 for each of them.</p>
<p>Is there any way to limit the scope of a PAT to a specific repository? or something that will fix this?</p>
<p>Thank you!</p>

**Comment by francisfuzz** [comment, marked as answer]
<blockquote>
<p><a class="mention" href="/mo-harris">@mo-harris</a> wrote:</p>
</blockquote>
<blockquote>
<p>Is there any way to limit the scope of a PAT to a specific repository? or something that will fix this?</p>
</blockquote>
<blockquote>
<p>Thank you!</p>
</blockquote>
<p>There’s not a way to limit the scope of a personal access token to a specific repository. We understand that you don’t want to grant applications permissions they don’t need, and these applications probably don’t even want to ask you for these permissions. In some cases, this might even prevent you from using an application.</p>
<p>It’s a very important topic and we’re glad you’re bringing it up with us. Providing more granular OAuth scopes is already the biggest blip on the API team’s radar and it’s something we’d love to do. However, we can’t make promises about if and when the scopes you wished for might be available – the API team is rolling out additional scopes as they are completed.</p>
<p>The best way to keep track of changes is to follow the API blog, and ping us if you have any questions about the available scopes:</p>
<ul>
<li><a href="https://developer.github.com/changes/" rel="nofollow noopener">https://developer.github.com/changes/</a></li>
</ul>
<p>Improving this situation about scopes is something the team has been working on. You might want to check out the GitHub Apps features:</p>
<ul>
<li><a href="https://developer.github.com/apps/" rel="nofollow noopener">https://developer.github.com/apps/</a></li>
</ul>
<p>GitHub Apps allow per-repository access and more finely-grained scopes. One appropach you could take is creating a GitHub App whose sole responsibility is to manage a single file in any of the repositories its installed to:</p>
<ul>
<li><a href="https://developer.github.com/apps/quickstart-guides/setting-up-your-development-environment/" rel="nofollow noopener">https://developer.github.com/apps/quickstart-guides/setting-up-your-development-environment/</a></li>
<li><a href="https://developer.github.com/v3/apps/permissions/#permission-on-single-file" rel="nofollow noopener">https://developer.github.com/v3/apps/permissions/#permission-on-single-file</a></li>
</ul>
<p>From there, you could install that GitHub App to the relevant repositories. Creating and using a GitHub App wouldn’t require you to buy another seat. In fact, you can create as many GitHub Apps as you need to perform these sort of programmatic and administrative operations that would have normally been delegated to another person on your team. We hope this helps!</p>

**Reply by joeflack4** [reply]
This is a very clear answer, thanks.

PAT's for single repositories in the future would be great. Would be even better if this could be created and managed within the repository's settings, rather than the PAT coming from a specific user.

My immediate use case is that we want to add a button to our existing web app that triggers a specific GitHub action.

**Comment by ghost** [comment]
<p>Thanks for the solution.</p>
<p>We ended up using GitHub App installation tokens with limited access to specific repositories.</p>
<p>From what I learnt from my testings, GitHub Packages don’t work with GitHub App installation tokens. Is that right or I’m missing something?</p>

**Comment by DingGGu** [comment]
<p>I can’t understand why it doesn’t have a feature yet.</p>
<p>I think it should add the per-repository limit feature as soon as possible.</p>
<p>Permissions are more intertwined than you might think.</p>
<p>Currently, It need to create an account on github to configure the above features.</p>
<ul>
<li>This called machine user in github.</li>
</ul>
<p>If I use Enterprise, I have to pay extra. Most importantly, changing the keys is inconvenient and this has a negative impact on security.</p>
<p>Also, having to form a private access token on the CI server or development server is terrible.</p>

**Comment by francisfuzz** [comment]
<aside class="quote no-group" data-username="mo-harris" data-post="3" data-topic="3129">
<div class="title">
<div class="quote-controls"></div>
 mo-harris:</div>
<blockquote>
<p>From what I learnt from my testings, GitHub Packages don’t work with GitHub App installation tokens. Is that right or I’m missing something?</p>
</blockquote>
</aside>
<p><a class="mention" href="/mo-harris">@mo-harris</a> - My apologies for not catching your question the first time around! :scream:</p>
<p>Our <a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/authentication-in-a-workflow#permissions-for-the-github_token">Authentication in a workflow’s section, Permissions for the <code>GITHUB_TOKEN</code></a> outlines each of the API endpoints a GitHub App can access. At this time of writing, packages are accessible by the <code>GITHUB_TOKEN</code> provided by GitHub Actions, where <code>read/write</code> access is granted on root repositories and <code>read</code> access is given on forked repositories.</p>
<hr>
<aside class="quote no-group" data-username="DingGGu" data-post="4" data-topic="3129">
<div class="title">
<div class="quote-controls"></div>
 DingGGu:</div>
<blockquote>
<p>I can’t understand why it doesn’t have a feature yet.</p>
<p>I think it should add the per-repository limit feature as soon as possible.</p>
</blockquote>
</aside>
<p>:wave: <a class="mention" href="/dingggu">@DingGGu</a> , hey there, and welcome to our GitHub Support Community! I acknowledge where you’re coming from and think the best place to share that feedback is with our product team <a href="https://support.github.com/contact/feedback">through this official feedback form</a>. If you’re interested in learning what our team is working on next, we suggest checking out our <a href="https://github.community/t/github-roadmap-faq/124978">GitHub Roadmap FAQs</a> for the latest.</p>
<aside class="quote no-group" data-username="DingGGu" data-post="4" data-topic="3129">
<div class="title">
<div class="quote-controls"></div>
 DingGGu:</div>
<blockquote>
<p>Permissions are more intertwined than you might think.</p>
<p>Currently, It need to create an account on github to configure the above features.</p>
<ul>
<li>This called machine user in github.</li>
</ul>
<p>If I use Enterprise, I have to pay extra. Most importantly, changing the keys is inconvenient and this has a negative impact on security.</p>
<p>Also, having to form a private access token on the CI server or development server is terrible.</p>
</blockquote>
</aside>
<p>I see where you’re coming from. Granular permissions came top of mind for the team who initially developed <a href="https://docs.github.com/en/free-pro-team@latest/developers/apps/about-apps">GitHub Apps</a>, and in my <a href="https://github.community/t/limiting-scope-of-a-pat-to-a-single-repository/3129/2">earlier reply</a>, I mentioned the approach of using GitHub Apps as an alternative from machine users. If you’ve tried GitHub Apps already and found that feature doesn’t meet your specific need, I’d be curious to learn more about what specific advantages you’ve found a machine user to have.</p>
<p>That additional context would help me understand where you’re coming from and I can check on our end to see if there are other alternatives that might be of interest!</p>

**Comment by DingGGu** [comment]
<p>Let me add some content to my question. One way to get authorize when pulling Docker images using GHCR is to use PAT.</p>
<p>Therefore, I need to register my Github PAT on the CI server or Kubernetes on Docker Credential.</p>
<p>I’m worried about this, and I’m looking for another way.</p>
<p>I think a great way to do this is to let Github create tokens that give you GHCR-per-container image-specific access.</p>
<p>In the current situation, I understand that even though the authority of Source and GHCR (which was added this time) are separated on Github, each team must have a different authority system, which creates N additional enterprise accounts and creates non-centralized tokens. It means you have to manage it individually.</p>
<p>Thanks for responding.</p>

**Comment by francisfuzz** [comment]
<p><a class="mention" href="/dingggu">@DingGGu</a> - Thanks for sharing that context with me. I’m going to look into this with one of my colleagues this week and see if there might be an alternative approach that could be useful. I’ll follow up by Friday with my findings (wherever I land).</p>

**Comment by ghost** [comment]
<p>If I may add a different use case.<br>
It would be super nice to be able to provide token access for <a href="http://raw.githubusercontent.com" rel="noopener nofollow ugc">raw.githubusercontent.com</a>.<br>
With tools like deno now supporting the execution of scripts via URL it would greatly simplify our deployment, and as a positive side effect, the load on github servers. Instead of having to do a pull each time we could just hit the CDN.</p>

**Comment by ProPuke** [comment]
<p>Alas 2 more years have passed and there still seems to be no way :frowning:</p>
<p>Just to raise the question again: Is there still no hope on getting more granular PAT control?</p>

**Comment by byrneh** [comment]
<p>It is currently listed in the “Future” section of GitHub’s Public Roadmap and not include in any of the next 4 quarters.<br>
<a href="https://github.com/github/roadmap/issues/184" rel="noopener nofollow ugc">Improved control over Personal Access Tokens (PATs)</a></p>

**Comment by pantelis-karamolegkos** [comment]
2 years later and no progress on this issue with huge security impact...?

**Comment by 1ur11** [comment]
Any update on topic ?

**Comment by Karlish-git** [comment]
For those that asked, when searching for solution, I stumbled on this thread. The issue linked by @byrneh (thanks) has been recently updated. 
At last, [Improved control over Personal Access Tokens (PATs)](https://github.com/github/roadmap/issues/184) is now on roadmap and target (at time ow writing) looks to be:
Status: Q4 2022 – Oct-Dec

**Reply by AllanOricil** [reply]
It is in Beta now

**Reply by andry81** [reply]
It's so fine grained that the ppl ask an external community for explanation what to select to use a simple push permission: https://stackoverflow.com/questions/75128935/what-permissions-should-i-choose-for-github-fine-grained-personal-tokens

**Comment by joeflack4** [comment]
Can we get this feature for organizations?

I don't see PATs as an option in the developer settings of my org. And when I got o my personal settings, it only shows repositories owned under my username's namespace. My [use case](https://github.com/orgs/community/discussions/21999#discussioncomment-6052968) applies to a public repository owned by an organization.

**Reply by miguetronic** [reply]
Hi @joeflack4  , have you found a solution or workaround for this since you posted?
